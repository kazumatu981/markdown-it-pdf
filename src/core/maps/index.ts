export * from './contents-map';
export * from './extension-map';
export * from './render-map';
export * from './contents-manager';
