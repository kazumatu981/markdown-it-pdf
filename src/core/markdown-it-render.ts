import MarkdownIt from 'markdown-it';
import fsPromises from 'fs/promises';

export interface Styles {
    internalUrls: string[];
    externalUrls: string[];
}

export class MarkdownItRender {
    private _md: MarkdownIt = new MarkdownIt();
    private _styles: Styles = {
        internalUrls: [],
        externalUrls: [],
    };

    /**
     * Adds the given URLs to the list of internal styles for this instance.
     *
     * @param {string[]} urls - The URLs of the styles to add.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addStyles(urls: string[]): this {
        this._styles.internalUrls = urls;
        return this;
    }

    /**
     * Sets the external style URLs of this instance.
     *
     * @param {string[]} urls - The URLs to set as external styles.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addExternalStyles(urls: string[]): this {
        this._styles.externalUrls = urls;
        return this;
    }

    /**
     * Uses the provided plugin with optional parameters.
     *
     * @param {MarkdownIt.PluginWithParams} plugin - The plugin to use.
     * @param {any[]} params - Optional parameters for the plugin.
     * @return {this} The MarkdownItRender instance for chaining.
     */
    public use(plugin: MarkdownIt.PluginWithParams, ...params: any[]): this {
        this._md.use(plugin, ...params);
        return this;
    }

    /**
     * Renders the given markdown string into HTML.
     *
     * @param {string} markdown - The markdown string to render.
     * @return {string} The rendered HTML string.
     */
    public render(markdown: string): string {
        const htmlBody = this._md.render(markdown);
        const styleTags = MarkdownItRender.generateStyleTags(
            this._styles.internalUrls,
            this._styles.externalUrls
        );
        return MarkdownItRender.htmlTemplate
            .replace(MarkdownItRender.styles, styleTags)
            .replace(MarkdownItRender.contents, htmlBody);
    }
    /**
     * Asynchronously renders a markdown file to HTML.
     *
     * @param {string} markdownFilePath - The path to the markdown file.
     * @return {Promise<string>} A promise that resolves to the rendered HTML.
     */
    public async renderFromFileAsync(
        markdownFilePath: string
    ): Promise<string> {
        const markdown = await fsPromises.readFile(markdownFilePath, 'utf8');
        return this.render(markdown);
    }

    //#region private members
    //#region Static Variables
    private static styles = '<!-- [[[STYLES]]] -->';
    private static contents = '<!-- [[[CONTENTS]]] -->';

    private static htmlTemplate = `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    ${MarkdownItRender.styles}
</head>
<body>
    ${MarkdownItRender.contents}
</body>`;
    //#endregion

    //#region Static Methods
    /**
     * Generates style tag based on the given URL.
     * @param {string} stylePath css file path
     * @returns {string} generated style tag
     */
    private static generateStyleTag(stylePath: string): string {
        return `<link rel="stylesheet" type="text/css" href="${stylePath}">`;
    }
    /**
     * Generates style tags based on the given internal and external URLs.
     *
     * @param {string[]} internalUrls - An optional array of internal URLs.
     * @param {string[]} externalUrls - An optional array of external URLs.
     * @return {string} The generated style tags as a string.
     */
    private static generateStyleTags(
        internalUrls?: string[],
        externalUrls?: string[]
    ): string {
        return (
            (internalUrls?.map(MarkdownItRender.generateStyleTag).join('\n') ??
                '') +
            (externalUrls?.map(MarkdownItRender.generateStyleTag).join('\n') ??
                '')
        );
    }
    //#endregion
    //#endregion
}
