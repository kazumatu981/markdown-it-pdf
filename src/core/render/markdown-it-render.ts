import MarkdownIt from 'markdown-it';
import fsPromises from 'fs/promises';
import { Logger } from '../../common/logger';
import { type Styles } from '../../common/configure';
import { type FileRender } from './file-render';

export class MarkdownItRender extends MarkdownIt implements FileRender {
    public _logger?: Logger;

    private _styles: Styles = {
        internalUrls: [],
        externalUrls: [],
    };

    /**
     * Adds the given URLs to the list of internal styles for this instance.
     *
     * @param {string[]} urls - The URLs of the styles to add.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addStyles(urls: string[]): this {
        this._styles.internalUrls.push(...urls);
        return this;
    }

    /**
     * Sets the external style URLs of this instance.
     *
     * @param {string[]} urls - The URLs to set as external styles.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addExternalStyles(urls: string[]): this {
        this._styles.externalUrls.push(...urls);
        return this;
    }

    /**
     * Renders the given markdown string into HTML.
     *
     * @param {string} markdown - The markdown string to render.
     * @return {string} The rendered HTML string.
     */
    public render(markdown: string): string {
        this._logger?.debug(`render() called.`);
        this._logger?.debug(`styles: %o`, this._styles);
        const htmlBody = super.render(markdown);
        const styleTags = MarkdownItRender.generateStyleTags(
            this._styles.internalUrls,
            this._styles.externalUrls
        );
        return MarkdownItRender.htmlTemplate
            .replace(MarkdownItRender.styles, styleTags)
            .replace(MarkdownItRender.contents, htmlBody);
    }
    /**
     * Asynchronously renders a markdown file to HTML.
     *
     * @param {string} markdownFilePath - The path to the markdown file.
     * @return {Promise<string>} A promise that resolves to the rendered HTML.
     */
    public async renderFromFile(markdownFilePath: string): Promise<string> {
        const markdown = await fsPromises.readFile(markdownFilePath, 'utf8');
        return this.render(markdown);
    }

    //#region private members
    //#region Static Variables
    private static styles = '<!-- [[[STYLES]]] -->';
    private static contents = '<!-- [[[CONTENTS]]] -->';

    private static htmlTemplate = `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    ${MarkdownItRender.styles}
</head>
<body>
    ${MarkdownItRender.contents}
</body>`;
    //#endregion

    //#region Static Methods
    /**
     * Generates style tag based on the given URL.
     * @param {string} stylePath css file path
     * @returns {string} generated style tag
     */
    private static generateStyleTag(stylePath: string): string {
        return `<link rel="stylesheet" type="text/css" href="${stylePath}">`;
    }
    /**
     * Generates style tags based on the given internal and external URLs.
     *
     * @param {string[]} internalUrls - An optional array of internal URLs.
     * @param {string[]} externalUrls - An optional array of external URLs.
     * @return {string} The generated style tags as a string.
     */
    private static generateStyleTags(
        internalUrls: string[],
        externalUrls: string[]
    ): string {
        return (
            internalUrls.map(MarkdownItRender.generateStyleTag).join('\n') +
            '\n' +
            externalUrls.map(MarkdownItRender.generateStyleTag).join('\n')
        );
    }
    //#endregion
    //#endregion
}
