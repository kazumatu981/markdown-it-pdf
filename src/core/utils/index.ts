export * from './http-helper';
export * from './path-resolver';
