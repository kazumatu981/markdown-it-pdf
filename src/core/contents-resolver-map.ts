import { MarkdownItRender } from './markdown-it-render';
import fsPromises from 'fs/promises';
export type ContentType = 'markdown' | 'style';
export type ContentsResolverFunction = (filePath: string) => Promise<string>;

export class ContentsResolver extends Map<
    ContentType,
    ContentsResolverFunction
> {
    private _render: MarkdownItRender;
    public constructor(render: MarkdownItRender) {
        super();
        this._render = render;
        this.set('markdown', async (filePath) => {
            return this._render.renderFromFileAsync(filePath);
        });
        this.set('style', async (filePath) => {
            return fsPromises.readFile(filePath, 'utf8');
        });
    }

    public static resolveMediaType(contentType: ContentType): string{
        switch (contentType) {
            case 'markdown':
                return 'text/html';
            case 'style':
                return 'text/css';
            default:
                return 'text/plain';
        }
    }
    public async resolve(type: ContentType, filePath: string): Promise<string> {
        const resolver = this.get(type);
        if (!resolver) {
            throw new Error(`Unsupported contents type: ${type}`);
        }
        return resolver(filePath);
    }
}
