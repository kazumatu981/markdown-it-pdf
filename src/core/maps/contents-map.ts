import { findFiles, filePathToUrl } from '../utils/path-resolver';
import { type RenderMap } from './render-map';
import { DefaultExtensionMap } from './extension-map';
import path from 'path';

//#region constants
const defaultRootDir = './';
/**
 * default contents map options
 */
const defaultContentsMapOptions = {
    recursive: true,
};
//#endregion

//#region types and interfaces

/**
 * An entity in the contents map
 */
export interface ContentsMapEntity {
    /**
     * The URL of the entity
     */
    url: string;
    /**
     * The resolver type of the entity
     */
    renderType: string;
    /**
     * The content type of the entity
     */
    contentType: string;
    /**
     * The path to the content of the entity
     */
    contentPath: string;
}

/**
 * An entity in the contents map that has been rendered
 */
export interface RenderedEntity extends ContentsMapEntity {
    /**
     * The contents of the entity
     */
    contents: string | Buffer;
}

/**
 * Options for the contents map
 */
export interface ContentsMapOptions {
    /**
     * Whether to recursively search for files.
     */
    recursive?: boolean;
}

//#endregion

//#region main class
/**
 * The contents map. It's a map of URLs to the contents.
 */
export class ContentsMap extends Map<string, ContentsMapEntity> {
    //#region private fields
    private _rootDir: string;
    private _renderMap: RenderMap;
    private _options?: ContentsMapOptions;
    //#endregion

    //#region public static methods

    /**
     * Creates an instance of the ContentsMap class with the provided resolver map and options.
     * Then refreshes the contents of the map and returns the instance.
     * @param {RenderMap} resolverMap - The resolver map used to resolve the URLs to the contents.
     * @param {string} rootDir - The root directory to start the search files from.
     * @param {ContentsMapOptions} options - The options for the contents map.
     * @returns {Promise<ContentsMap>} A promise that resolves to the created instance of the ContentsMap class.
     */
    public static async createInstance(
        resolverMap: RenderMap,
        rootDir?: string,
        options?: ContentsMapOptions
    ): Promise<ContentsMap> {
        // create the instance
        const theInstance = new ContentsMap(resolverMap, rootDir, options);

        // refresh the contents of the map
        await theInstance.refresh();

        // return the created instance
        return theInstance;
    }
    //#endregion

    //#region constructor

    /**
     * Creates a new instance of the ContentsMap class.
     * @param {RenderMap} resolverMap - The resolver map used to resolve the URLs to the resolver for the contents.
     * @param {string} rootDir - The root directory to start the search files from.
     * @param {ContentsMapOptions} options - The options for the contents map.
     */
    private constructor(
        resolverMap: RenderMap,
        rootDir?: string,
        options?: ContentsMapOptions
    ) {
        // Call the constructor of the parent class (Map)
        super();

        // Initialize the resolver map
        this._renderMap = resolverMap;

        // Initialize the root directory
        this._rootDir = rootDir ?? defaultRootDir;

        // Initialize the options for the contents map
        this._options = options;
    }

    //#endregion

    //#region public properties
    public get renderMap(): RenderMap {
        return this._renderMap;
    }
    public set renderMap(renderMap: RenderMap) {
        this._renderMap = renderMap;
    }
    //#endregion

    //#region public methods
    /**
     * Refreshes the contents of the map by clearing the current contents and
     * re-populating it with the contents found in the root directory using the
     * provided options.
     * @returns {Promise<void>} A promise that resolves when the contents are
     * refreshed.
     */
    public async refresh(): Promise<void> {
        // Clear the current contents of the map
        this.clear();

        // Find the contents of the root directory using the provided options
        const contents = await findFiles<ContentsMapEntity>(
            this._rootDir,
            this._options?.recursive ?? defaultContentsMapOptions.recursive,
            DefaultExtensionMap.isSupported.bind(DefaultExtensionMap),
            this.generateContentMapEntity.bind(this)
        );

        // Populate the map with the found contents
        contents.forEach((entry) => {
            this.set(entry.url, entry);
        });
    }

    /**
     * Returns an array of entity URLs based on the provided renderType.
     * If renderType is not provided, all entity URLs are returned.
     * @param {string} renderType - The render type of the entities to retrieve.
     * @returns {string[]} An array of entity URLs that match the provided renderType,
     * or all entity URLs if renderType is not provided.
     */
    public getEntityPaths(renderType?: string): string[] {
        // If renderType is provided, filter the entity URLs based on the render type
        return renderType
            ? [...this.keys()].filter((url) => {
                  // Get the entity associated with the URL
                  const entity = this.get(url) as ContentsMapEntity;
                  // Return true if the entity's render type matches the provided renderType
                  return entity.renderType === renderType;
              })
            : [...this.keys()];
    }

    /**
     * Renders the entity associated with the provided filePath.
     * If the entity is not found, it returns undefined.
     * @param {string} filePath - The path of the file to render.
     * @returns {Promise<RenderedEntity | undefined>} - A promise that resolves to the rendered entity,
     * or undefined if the entity is not found.
     */
    public async render(filePath: string): Promise<RenderedEntity | undefined> {
        // Get the entity associated with the provided filePath.
        const entity = this.get(filePath);

        // If the entity is not found, return undefined.
        if (!entity) {
            return;
        }

        // Get the render for the renderType of the entity.
        const contentRender = this._renderMap.getRender(entity.renderType);

        // Render the contents of the entity using the render.
        const contents = await contentRender.renderFromFile(entity.contentPath);

        // Return the rendered entity with the contents.
        return { ...entity, contents };
    }

    //#endregion

    //#region private methods
    /**
     * Generates a content map entity for the provided file path.
     * @param {string} filePath - The path of the file.
     * @returns {ContentsMapEntity} The generated content map entity.
     */
    private generateContentMapEntity(filePath: string): ContentsMapEntity {
        // Get the resolver information for the file extension.
        const resolver = DefaultExtensionMap.getTypeInfo(
            path.extname(filePath)
        );

        // Generate the content map entity with the provided information.
        return {
            // Generate the URL for the file path.
            url: filePathToUrl(this._rootDir, filePath),
            // Get the resolver type from the resolver information.
            renderType: resolver.renderType,
            // Get the content type from the resolver information.
            // If the resolved content type is available, use it instead.
            contentType: resolver.resolvedContentType ?? resolver.contentType,
            // Set the content path to the provided file path.
            contentPath: filePath,
        };
    }

    //#endregion
}
//#endregion
