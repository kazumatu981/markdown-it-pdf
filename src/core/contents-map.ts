import { type MarkdownItRender } from './markdown-it-render';
import { type ContentType, ContentsResolver } from './contents-resolver-map';
export interface ContentsMapEntity {
    type: ContentType;
    contentPath: string;
}
export interface RenderedEntity extends ContentsMapEntity {
    mediaType: string;
    contents: string;
}

export class ContentsMap extends Map<string, ContentsMapEntity> {
    private _resolver: ContentsResolver;
    public constructor(render: MarkdownItRender) {
        super();
        this._resolver = new ContentsResolver(render);
    }

    public async render(filePath: string): Promise<RenderedEntity | undefined> {
        const entity = this.get(filePath);
        if (!entity) {
            return;
        }
        const contents = await this._resolver.resolve(
            entity.type,
            entity.contentPath
        );
        const mediaType = ContentsResolver.resolveMediaType(entity.type);
        return { ...entity, contents, mediaType };
    }
}
