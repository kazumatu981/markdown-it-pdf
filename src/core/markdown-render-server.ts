import http from 'http';

import { type ContentsMap, RenderedEntity } from './contents-map';

export class MarkdownRenderServer {
    private _server: http.Server;
    private _map?: ContentsMap;
    constructor() {
        this._server = http.createServer(this.serverListener.bind(this));
    }

    public map(map: ContentsMap): this {
        this._map = map;
        return this;
    }

    public serverListener(
        req: http.IncomingMessage,
        res: http.ServerResponse
    ): void {
        if (!this._map) {
            this.writeNotFound(res);
            return;
        }
        const filePath = req.url ?? '';
        this._map
            .render(filePath)
            .then((entity) => {
                if (!entity) {
                    this.writeNotFound(res);
                    return;
                }
                this.writeEntity(res, entity);
            })
            .catch((err) => {
                console.error(err);
                this.writeNotFound(res);
            });
    }
    public listen(port: number): void {
        this._server.listen(port);
    }
    public close(): void {
        this._server.close();
    }

    public writeNotFound(res: http.ServerResponse): void {
        res.statusCode = 404;
        res.appendHeader('Content-Type', 'text/plain');
        res.write('Not Found');
        res.end();
    }
    public writeEntity(res: http.ServerResponse, entity: RenderedEntity): void {
        res.statusCode = 200;
        res.appendHeader('Content-Type', entity.mediaType);
        res.write(entity.contents);
        res.end();
    }
}
