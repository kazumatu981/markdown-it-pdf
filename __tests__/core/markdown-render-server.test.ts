import { jest, expect, describe, it } from '@jest/globals';
import { MarkdownRenderServer } from '../../src/core/markdown-render-server';
import { ContentsMap } from '../../src/core/contents-map';
import { MarkdownItRender } from '../../src/core/markdown-it-render';
import http from 'http';
import { mockingTestDir, unmockingTestDir } from '../utils/test-dir';

function readFromServer(url: string): Promise<string> {
    return new Promise((resolve, reject) => {
        http.get(url, (res) => {
            let data = '';
            res.on('data', (chunk) => {
                data += chunk;
            });
            res.on('end', () => {
                resolve(data);
            });
        });
    });
}

describe('CoreLibrary Unit Tests - MarkdownRenderServer', () => {
    it('Basic Test', async () => {
        mockingTestDir();
        const render = new MarkdownItRender();

        const contentsMap = new ContentsMap(render);
        contentsMap.set('/test/test.md', {
            type: 'markdown',
            contentPath: 'test/test.md',
        });

        const server = new MarkdownRenderServer();
        server.map(contentsMap).listen(3000);

        const data = await readFromServer('http://localhost:3000/test/test.md');

        server.close();
        unmockingTestDir();
        expect(data).toMatchSnapshot();
    });
});
