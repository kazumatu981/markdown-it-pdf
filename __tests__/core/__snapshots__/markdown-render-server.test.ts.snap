// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreLibrary Unit Tests - MarkdownRenderServer Basic Test 1`] = `
"
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    
</head>
<body>
    <h1>test</h1>
<p>hello world</p>

</body>"
`;
